for_curators/sample_local_dataset_upload_dev.py.ipynb

```diff
@@ -0,0 +1,160 @@
+{
+ "cells": [
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "id": "92063ed7",
+   "metadata": {
+    "scrolled": false
+   },
+   "outputs": [],
+   "source": [
+    "import boto3\n",
+    "import os\n",
+    "import requests\n",
+    "import threading\n",
+    "from botocore.credentials import RefreshableCredentials\n",
+    "from botocore.session import get_session\n",
+    "from datetime import datetime, timezone\n",
+    "\n",
+    "\n",
+    "# This script performs the upload of a local datafile to a given Collection (as identified by its Collection uuid),\n",
+    "# where the datafile becomes a Dataset accessible via the Data Portal UI. In order to use this script, you must have\n",
+    "# a Curation API key, you must know the id of the Collection to which you wish to upload the datafile, and you must \n",
+    "# decide upon a string tag (the `curator_tag`) to use to uniquely identify the resultant Dataset within this \n",
+    "# Collection going forward. Uploads to a tag that has not been used yet in the given Collection will result in a\n",
+    "# new Dataset being created. Uploads to a tag for which there already exists a Dataset in the given Collection will \n",
+    "# result in the existing Dataset being replaced by the new Dataset created from the datafile that you are uploading.\n",
```

Comment:
What is the advantage of having this presented as a notebook rather than just a script? This presentation is not taking advantage of common notebook features, like markdown for headers and comments, and cell-per-step structuring. I'm just thinking that the code might actually be easier adapted if it can be directly copy & pasted & edited. Are do we know curators _are_ going to want to use this in a notebook env?

Reply:
I went with a notebook because that was Brian's preferred format; the curators tend to use notebooks to execute their code.

> This presentation is not taking advantage of common notebook features, like markdown for headers and comments,

I did not know about this! Will look into it, thank you

>  cell-per-step structuring

I know about this but I was iffy on whether it was actually useful given the fairly contained purpose of the script that this notebook has. I _guess_ we could break up into separate cells along the lines that I've already laid out with the commented headers (3 steps), but the curator can also do that themselves if they're interested in doing so; if they actually find that they need things broken up, that means they necessarily have sufficient capability to do so themselves, and therefore they can.

Reply:
A notebook is usually optimized for ease of reading and understanding, which begs for having first-class cells containing headers and documentation.  TileDB has some good examples: https://cloud.tiledb.com/notebooks/details/TileDB-Inc/tutorials_dense_basics/preview

Reply:
I just think those that are familiar with notebooks will look at this one and question why it ain't pretty. 😄 Functionally, it's not a problem.  I'm biased towards scripts, personally.

Reply:
@Bento007 @atolopko-czi I definitely want to hear your thoughts on the last part when you get a chance. ☝️

Reply:
Ok, that's a great point in terms of what they're going to be expecting. Taking a look at the example you linked to...

Reply:
> If sticking with a notebook format, it's not uncommon to include the pip install requirements as a cell and use `%pip` magic command (https://ipython.readthedocs.io/en/stable/interactive/magics.html#magic-pip).

I'm actually not a huge fan of this since you can't use it for installing jupyter/notebook packages themselves; those must already be installed.

I'm just bundling all reqs into `requirements.txt`.

Reply:
I have added proper markdown, though! 😄